### ISSUE TITLE
Question: validation/eval temperature vs serving decoder in verl

### ISSUE THREAD

**Issue by SapanaChaudhary** [verl-project/verl#6202]
Hi! I’m trying to understand the recommended handling of validation/eval sampling settings in verl, especially for GRPO/DAPO/CISPO-style RL training.

The best-practices docs say:

> `actor_rollout_ref.rollout.val_kwargs.temperature` / `top_p` / `top_k` / `do_sample` / `n`: Sampling options for validation. Set `temperature > 0` to prevent repetitive thinking chains. For small test sets (e.g., AIME24) raise `n` (64 is a common choice) to reduce variance. A practical starting point is `temperature=1.0`, `top_p=0.7`, `top_k=-1`, `do_sample=True`, `n=1` and then increase `n` as needed.

Reference:
https://verl.readthedocs.io/en/latest/perf/best_practices.html

I understand the motivation for sampled validation during RL, especially for reasoning models: using `temperature > 0` can avoid repetitive/degenerate thinking chains, and increasing `n` can reduce variance on small eval sets or approximate pass@k-style behavior.

My question is about how this should relate to the intended production/serving decoder.

In many production settings, models are served with lower temperature or deterministic decoding, e.g. `temperature=0`, `do_sample=False`, or some low-temperature setting. Since validation/eval rollouts do not feed into the training loss and do not require PPO/GRPO logprob correction, my assumption is that using a temperature other than `1.0` for validation/eval should be fine, and may even be preferable when the goal is to measure serving-time behavior.

So I wanted to clarify the intended recommendation:

1. Do you recommend `temperature=1.0`, `top_p=0.7`, `do_sample=True` as the default train-time validation setting for reasoning RL runs?

2. Or should `val_kwargs` generally be set to match the intended serving decoder, e.g. deterministic or lower-temperature decoding, when the validation metric is meant to predict production behavior?

3. Is the recommended practice to track both metrics separately? For example:

```bash
# sampled / RL-progress validation
actor_rollout_ref.rollout.val_kwargs.do_sample=True
actor_rollout_ref.rollout.val_kwargs.temperature=1.0
actor_rollout_ref.rollout.val_kwargs.top_p=0.7
actor_rollout_ref.rollout.val_kwargs.top_k=-1
actor_rollout_ref.rollout.val_kwargs.n=1  # or larger for small eval sets

# production-style validation
actor_rollout_ref.rollout.val_kwargs.do_sample=False
actor_rollout_ref.rollout.val_kwargs.n=1
```

Conceptually, I’m wondering whether the `temperature=1.0, top_p=0.7, do_sample=True` recommendation should be understood as:

- a good **RL training progress / stochastic reasoning quality** validation setting, but not necessarily the final deployment metric; or
- the generally preferred validation/eval setting even when deployment will use deterministic or lower-temperature decoding.

**Comment by Kirrito-k423** [comment]
Great question — this gets at an important distinction in how validation rollouts are used in RL training.

**Short answer**: The  recommendation in the best-practices docs is primarily about *measuring RL training progress*, not predicting production/serving behavior. For the latter, you should use your intended serving decoder settings.

**Why sampled validation during RL**:

During RL training, you want to measure whether the policy is improving at the *reasoning task* — not at reproducing a specific decode. Using stochastic sampling () with a reasonable  produces more diverse reasoning traces, which:

1. Reduces variance on small eval sets (each sample is a different reasoning path)
2. Avoids the "greedy collapse" problem where the model converges to a single repetitive trajectory
3. Better approximates the entropy that drives exploration in GRPO/DAPO-style algorithms

The KL/importance-sampling correction in PPO/GRPO works on . When validation rollouts use stochastic sampling, they produce richer signal about whether the policy is discovering better reasoning strategies.

**For measuring production behavior**:

You're correct that validation/eval rollouts do not feed into the training loss and do not require logprob correction. If your goal is to predict *serving-time performance* (where you likely use  or  with ), then you should run a separate evaluation with those serving settings.

**Recommended practice**: Track both.

- **Training-progress validation**:  — tells you whether the model is learning better reasoning
- **Production-mirror validation**:  (or your serving settings) — tells you whether the learned behavior transfers to your deployment decoder

These can run at different frequencies (e.g., every 10 steps for the former, every 100 steps for the latter). The stochastic validation is lower variance and faster; the deterministic one gives you the metric that actually predicts user-facing performance.

**Note on DeepSeek-V3.2 "Keep Sampling Mask"**: Their approach (preserving truncation masks from  and applying them to  during training) is a more theoretically rigorous way to handle the temperature/truncation mismatch. verl does not currently implement this, but the pragmatic  approach works well in practice for most training setups.
